text/profiles/0237-profiles-data-model.md

```diff
@@ -0,0 +1,1015 @@
+                attribute_set_indices:
+                  - 1
+                  - 1
+                link_indices:
+                  - 1
+                  - 0
+                stacktrace_indices:
+                  - 0
+                  - 1
+                timestamps:
+                  - 1.687841528e+15
+                  - 0
+                values:
+                  - 100
+                  - 200
+            stacktraces:
+              - location_indices:
+                  - 0
+                  - 1
+                  - 2
+              - location_indices:
+                  - 0
+                  - 1
+            string_table:
+              - ""
+              - foo
+              - bar
+              - baz
+              - cpu
+              - samples
+            attributes: null
+            profile_id: 0102030405060708090a0b0c0d0e0f10
+        scope:
+          attributes: null
+```
+
+### Notable differences compared to other signals
+
+Due to the increased performance requirements associated with profiles signal, here are some notable differences between profiles signal and other signals.
+
+#### Relationships between messages
+
+There are two main ways relationships between messages are represented:
+
+* by embedding a message into another message (standard protobuf way)
+* by referencing a message by index (similar to how it's done in pprof)
+
+Profiling signal is different from most other ones in that we use the referencing technique a lot to represent relationships between messages where there is a lot of duplication happening. This allows to reduce the size of the resulting protobuf payload and the number of objects that need to be allocated to parse such payload.
+
+This pseudocode illustrates the conceptual difference between the two approaches. Note that this example is simplified for the sake of clarity:
+
+```json
+// denormalized
+"samples": [
+  {
+    "stacktrace": ["foo", "bar"],
+    "value": 100,
+    "attribute_set": {
+      "endpoint": "/v1/users"
+    }
+  }, {
+    "stacktrace": ["foo", "bar", "baz"],
+    "value": 200,
+    "attribute_set": {
+      "endpoint": "/v1/users"
+    }
+  }
+],
+
+// normalized
+"attribute_sets": [
+  {
+    "endpoint": "/v1/users"
+  }
+],
+"samples": [
+  {
+    "stacktrace": ["foo", "bar"],
+    "value": 100,
+    "attribute_set_index": 0
+  }, {
+    "stacktrace": ["foo", "bar", "baz"],
+    "value": 200,
+    "attribute_set_index": 0
+  }
+],
+```
+
+Explanation: because multiple samples have the same attributes, we can store them in a separate table and reference them by index. This reduces the size of the resulting protobuf payload and the number of objects that need to be allocated to parse such payload.
+
+Benchmarking shows that this approach is significantly more efficient in terms of CPU utilization, memory consumption and size of the resulting protobuf payload. See [Prior art and alternatives](#prior-art-and-alternatives) for more details.
```

Comment:
I already commented elsewhere but want to emphasize one more time. My past experiments with Protobuf, string lookup tables and compression has shown me that string lookup table can result in an _increase_ of wire size in the compressed form (depending on the payload composition). The GZIP/ZSTD compression is very efficient in eliminating duplicate strings which negates the need to pre-optimize the payload with lookup tables and in fact usage of lookup tables is in some cases penalized by GZIP/ZSTD compressors who don't necessarily compress string indices represented as varints well.

A varint may also occupy more bytes than the string (as an extreme case imagine a short string "a", requiring only 2 bytes on the wire and this string corresponding to a large index in the the lookup table requiring more than 2 bytes on the wire), so this may require elaborate techniques to optimize the lookup table ordering by string length (and this can be quite costly during encoding).

Before we commit to usage of string lookups (and other lookups) I would want to see a comparison benchmark showing uncompressed and compressed wire sizes with GZIP and ZSTD for the versions of the profiling proto with the particular field using a direct string value vs string lookup. This ideally needs to be done on realistic production-like data and ideally one field at a time. I know this can be tedious but that's nature of the work.

This does not necessarily need to be done as part of the OTEP but it is necessary before it becomes part of OTLP in the proto repository.

Reply:
> I already commented elsewhere but want to emphasize one more time. My past experiments with Protobuf, string lookup tables and compression has shown me that string lookup table can result in an _increase_ of wire size in the compressed form (depending on the payload composition). The GZIP/ZSTD compression is very efficient in eliminating duplicate strings which negates the need to pre-optimize the payload with lookup tables and in fact usage of lookup tables is in some cases penalized by GZIP/ZSTD compressors who don't necessarily compress string indices represented as varints well.
> 
> A varint may also occupy more bytes than the string (as an extreme case imagine a short string "a", requiring only 2 bytes on the wire and this string corresponding to a large index in the the lookup table requiring more than 2 bytes on the wire), so this may require elaborate techniques to optimize the lookup table ordering by string length (and this can be quite costly during encoding).
> 
> Before we commit to usage of string lookups (and other lookups) I would want to see a comparison benchmark showing uncompressed and compressed wire sizes with GZIP and ZSTD for the versions of the profiling proto with the particular field using a direct string value vs string lookup. This ideally needs to be done on realistic production-like data and ideally one field at a time. I know this can be tedious but that's nature of the work.
> 
> This does not necessarily need to be done as part of the OTEP but it is necessary before it becomes part of OTLP in the proto repository.

@tigrannajaryan Compressed size is one thing (and should be compared, yes), but smaller uncompressed size has its benefits since the profile needs to be held uncompressed at some points on the client and server. Also, string interning gives the additional benefit as the profile is processed in a GC language like Go or Java since the deserialized string will get naturally shared in case of denormalized memory representation.